model/generate.go

```diff
@@ -107,72 +106,64 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: fmt.Sprintf("unable to find version %s", t.Version)}
 	}
-	if v.Config == "" {
-		return nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: fmt.Sprintf("unable to find config string for version %s", t.Version)}
-	}
-	p := &Project{}
-	if err = LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
-		return nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "error reading project yaml for version %s", t.Version).Error()}
+	p, pp, err := LoadProjectForVersion(v, t.Project, false)
+	if err != nil {
+		return nil, nil, nil, nil, nil,
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "error getting project for version %s", t.Version).Error()}
 	}
 
 	// Cache project data in maps for quick lookup
 	cachedProject := cacheProjectData(p)
 
 	// Validate generated project against original project.
 	if err = g.validateGeneratedProject(p, cachedProject); err != nil {
-		return nil, nil, nil, nil,
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "generated project is invalid").Error()}
 	}
 
-	newConfig, err := g.addGeneratedProjectToConfig(v.Config, cachedProject)
+	newPP, newConfig, err := g.addGeneratedProjectToConfig(pp, v.Config, cachedProject)
 	if err != nil {
-		return nil, nil, nil, nil,
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error creating config from generated config").Error()}
 	}
+	newPP.Id = v.Id
 	v.Config = newConfig
-	if err := LoadProjectInto([]byte(v.Config), t.Project, p); err != nil {
-		return nil, nil, nil, nil,
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error reading project yaml").Error()}
+	p, err = translateProject(newPP)
+	if err != nil {
+		return nil, nil, nil, nil, nil,
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrap(err, "error translating project").Error()}
 	}
-	return p, v, t, &cachedProject, nil
+	return p, newPP, v, t, &cachedProject, nil
 }
 
-func (g *GeneratedProject) Save(ctx context.Context, p *Project, v *Version, t *task.Task, pm *projectMaps) error {
-	query := bson.M{
-		VersionIdKey:           v.Id,
-		VersionConfigNumberKey: v.ConfigUpdateNumber,
-	}
-	update := bson.M{
-		"$set": bson.M{VersionConfigKey: v.Config},
-		"$inc": bson.M{VersionConfigNumberKey: 1},
-	}
-
-	err := VersionUpdateOne(query, update)
-	if err != nil {
+func (g *GeneratedProject) Save(ctx context.Context, p *Project, pp *ParserProject, v *Version, t *task.Task, pm *projectMaps) error {
+	if err := VersionUpdateConfig(v.Id, v.Config, v.ConfigUpdateNumber); err != nil {
 		return errors.Wrapf(err, "error updating version %s", v.Id)
 	}
+	v.ConfigUpdateNumber += 1
```

Comment:
Update: we tried this, it seems to be impacting the repotracker, we're not inclined to push it because this code is already in production and isn't racing.